### PR TITLE
use cu instead of mcrz-mcry-mcrz in add_control for gate u

### PR DESCRIPTION
**Pull request by hhorii**
### Summary

use `cu` instead of `mcrz`-`mcry`-`mcrz` in add_control for gate `u`

### Details and comments

Currently, 'u' in a circuit `circ` is converted to `mcrz`-`mcry`-`mcrz` when `circ.control()` is called. `cu` is better for simplicity.

### CODE REVIEW

**Comment by ewinston** [comment]
I believe the reason `cu` wasn't used was because it had a higher gate count than this approach at the time. The `<=` tests which are failing were meant to make sure alternative approaches don't increase the gate count.

**Comment by Cryoris** [comment]
I think in the one qubit case `CU` is more efficient since it only uses two CNOTs. Right now this is rather inefficient:
```python
>>> qc = QuantumCircuit(1)
>>> qc.p(0.2, 0)
>>> qc.u(1,2,3,0)

q0_0: ──────■──────
      ┌─────┴─────┐
 q_0: ┤ circuit10 ├
      └───────────┘
>>> qc.control().decompose().draw()
                ┌────────┐                                                                      ┌──────┐
q1_0: ────■─────┤ P(1.5) ├──■───────────────────■───────────────────────────■────────────────■──┤ P(1) ├──■─────────────────■──────────────
      ┌───┴────┐├────────┤┌─┴─┐┌─────────────┐┌─┴─┐┌──────────┐┌─────────┐┌─┴─┐┌──────────┐┌─┴─┐├──────┤┌─┴─┐┌───────────┐┌─┴─┐┌──────────┐
 q_0: ┤ P(0.2) ├┤ P(1.5) ├┤ X ├┤ U(0,0,-1.5) ├┤ X ├┤ U(0,0,0) ├┤ RY(0.5) ├┤ X ├┤ RY(-0.5) ├┤ X ├┤ P(1) ├┤ X ├┤ U(0,0,-1) ├┤ X ├┤ U(0,0,0) ├
      └────────┘└────────┘└───┘└─────────────┘└───┘└──────────┘└─────────┘└───┘└──────────┘└───┘└──────┘└───┘└───────────┘└───┘└──────────┘
```

@hhorii Can you add a test that covers this? E.g. the simple case above would already be fine I think.

**Comment by hhorii** [comment]
@Cryoris I added two test cases for `cp` and `cu` from controlled `u1` and `u`, respectively. I used `decompose()` twice instead of `Unroller(['cx', 'u', 'cp', 'cu']).run()` because `Unroller` may generate `cp` and `cu` by aggregating gates, in the future.

File: test/python/circuit/test_controlled_gate.py [hunk 1]
```diff
@@ -82,6 +82,16 @@ def test_controlled_u1(self):
         theta = 0.5
         self.assertEqual(U1Gate(theta).control(), CU1Gate(theta))
 
+        circ = QuantumCircuit(1)
+        circ.append(U1Gate(theta), circ.qregs[0])
+
+        unroller = Unroller(['cx', 'u', 'u1'])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py:88]:
```suggestion
        unroller = Unroller(['cx', 'u', 'p'])
```

File: test/python/circuit/test_controlled_gate.py [hunk 2]
```diff
@@ -109,6 +119,17 @@ def test_controlled_u3(self):
         self.assertEqual(U3Gate(theta, phi, lamb).control(),
                          CU3Gate(theta, phi, lamb))
 
+        circ = QuantumCircuit(1)
+        circ.append(U3Gate(theta, phi, lamb), circ.qregs[0])
+
+        unroller = Unroller(['cx', 'u', 'u1'])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py:125]:
```suggestion
        unroller = Unroller(['cx', 'u', 'p'])
```

**Review by Cryoris — APPROVED** [review]
LGTM thanks for the improvement 👍

**Pull request merged**